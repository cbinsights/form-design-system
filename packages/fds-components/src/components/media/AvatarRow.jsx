import React from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';
import Avatar from '@material-ui/core/Avatar';

/**
 * @param {Array} list original member list
 * @param {Number} max maximum avatars to render
 * @returns {Object} object containing overflow count and list to render
 */
export const getRenderConfig = (list, max) => {
  const renderList = list.slice(0, max);
  return {
    renderList,
    count: list.length - renderList.length,
  };
};

/**
 * @param {Object} props react props
 * @returns {ReactElement}
 */
const AvatarRow = (props) => {
  const { memberList, size, max, className } = props;
  const renderConfig = getRenderConfig(memberList, max);

  const classNames = cx('avatarRow', `avatarRow--${size}`, className);

  const fontClass = cx({
    'typemod--tiny': size === 's',
    'typemod--small': size === 'm',
    'typemod--xxlarge': size === 'xl',
  });

  return (
    <ul className={classNames}>
      {/* avatars */}
      {renderConfig.renderList.map((member, i) => (
        <li className="avatarRow-item" key={`${member.name.replace(/\s/g, '')}-${i}`}>
          {member.src ? (
            <Avatar alt={member.name} src={member.src} className="avatarRow-avatar" />
          ) : (
            <Avatar alt={member.name} className={cx('avatarRow-avatar', fontClass)}>
              {member.name.charAt(0).toUpperCase()}
            </Avatar>
          )}
        </li>
      ))}

      {/* overflow count bubble */}
      {renderConfig.count > 0 && (
        <li className="avatarRow-item">
          <Avatar className="avatarRow-avatar avatarRow-avatar--count">
            <span className={cx('typemod--book', fontClass)}>+{renderConfig.count}</span>
          </Avatar>
        </li>
      )}
    </ul>
  );
};

AvatarRow.defaultProps = {
  max: 3,
  size: 'm',
};

AvatarRow.propTypes = {
  /** List of members to render - `[ {name, src}, ... ]` */
  memberList: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      src: PropTypes.string,
    })
  ).isRequired,

  /** Size of avatars */
  size: PropTypes.oneOf(['s', 'm', 'l', 'xl']),

  /** Maximum number of avatars to render */
  max: PropTypes.number,

  /** Classes to pass to AvatarRow wrapper */
  className: PropTypes.string,
};

export default AvatarRow;
