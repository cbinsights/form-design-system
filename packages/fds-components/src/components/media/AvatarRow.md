
### AvatarRow
Render an overlapping list of avatars with an overflow count. The maxmimum number
of avatars to render can be configred with the `max` prop.

#### Default AvatarRow

```js
<AvatarRow
  memberList={[
    { name: "member mcmemberton", src: "https://placeimg.com/80/81/animals" },
    { name: "member mcmemberton", src: "https://placeimg.com/81/80/animals" },
    { name: "member mcmemberton", src: "https://placeimg.com/82/80/animals" },
    { name: "member mcmemberton" },
    { name: "member mcmemberton", src: "https://placeimg.com/83/80/animals" },
    { name: "member mcmemberton", src: "https://placeimg.com/80/83/animals" },
    { name: "member mcmemberton", src: "https://placeimg.com/84/80/animals" },
    { name: "member mcmemberton", src: "https://placeimg.com/80/84/animals" },
  ]}
/>
```

#### AvatarRow with higher `max` and larger size

```js
<AvatarRow
  max={6}
  size='l'
  memberList={[
    { name: "member mcmemberton", src: "https://placeimg.com/80/81/animals" },
    { name: "member mcmemberton", src: "https://placeimg.com/81/80/animals" },
    { name: "member mcmemberton", src: "https://placeimg.com/82/80/animals" },
    { name: "member mcmemberton" },
    { name: "member mcmemberton", src: "https://placeimg.com/83/80/animals" },
    { name: "member mcmemberton", src: "https://placeimg.com/80/83/animals" },
    { name: "member mcmemberton", src: "https://placeimg.com/84/80/animals" },
    { name: "member mcmemberton", src: "https://placeimg.com/80/84/animals" },
  ]}
/>
```

#### `max` exceeds list length

```js
<AvatarRow
  max={99}
  memberList={[
    { name: "member mcmemberton", src: "https://placeimg.com/80/81/animals" },
    { name: "member mcmemberton", src: "https://placeimg.com/81/80/animals" },
    { name: "member mcmemberton", src: "https://placeimg.com/82/80/animals" },
    { name: "member mcmemberton" },
    { name: "member mcmemberton", src: "https://placeimg.com/83/80/animals" },
    { name: "member mcmemberton", src: "https://placeimg.com/80/83/animals" },
    { name: "member mcmemberton", src: "https://placeimg.com/84/80/animals" },
    { name: "member mcmemberton", src: "https://placeimg.com/80/84/animals" },
  ]}
/>
```
