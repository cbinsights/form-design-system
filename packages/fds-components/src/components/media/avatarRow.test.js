import React from 'react';
import { shallow } from 'enzyme';

import AvatarRow, { getRenderConfig } from './AvatarRow';

const renderComponent = (props) => shallow(<AvatarRow {...props} />);

const MOCK_PROPS = {
  memberList: [
    { name: 'MOCK NAME', src: 'https://placeimg.com/40/41/animals' },
    { name: 'MOCK NAME', src: 'https://placeimg.com/40/41/animals' },
    { name: 'MOCK NAME', src: 'https://placeimg.com/40/41/animals' },
    { name: 'MOCK NAME', src: 'https://placeimg.com/40/41/animals' },
    { name: 'MOCK NAME', src: 'https://placeimg.com/40/41/animals' },
  ]
};

describe('AvatarRow component', () => {

  it('matches snapshot', () => {
    const component = renderComponent(MOCK_PROPS);
    expect(component).toMatchSnapshot();
  });

  it('getRenderConfig returns correct render list and count', () => {
    const actual = getRenderConfig(MOCK_PROPS.memberList, 4);
    expect(actual.renderList.length).toBe(4);
    expect(actual.count).toBe(1);
  });
});
