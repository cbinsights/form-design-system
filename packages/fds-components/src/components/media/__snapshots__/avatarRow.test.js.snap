// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarRow component matches snapshot 1`] = `
<ul
  className="avatarRow avatarRow--m"
>
  <li
    className="avatarRow-item"
    key="MOCKNAME-0"
  >
    <WithStyles(Avatar)
      alt="MOCK NAME"
      className="avatarRow-avatar"
      src="https://placeimg.com/40/41/animals"
    />
  </li>
  <li
    className="avatarRow-item"
    key="MOCKNAME-1"
  >
    <WithStyles(Avatar)
      alt="MOCK NAME"
      className="avatarRow-avatar"
      src="https://placeimg.com/40/41/animals"
    />
  </li>
  <li
    className="avatarRow-item"
    key="MOCKNAME-2"
  >
    <WithStyles(Avatar)
      alt="MOCK NAME"
      className="avatarRow-avatar"
      src="https://placeimg.com/40/41/animals"
    />
  </li>
  <li
    className="avatarRow-item"
  >
    <WithStyles(Avatar)
      className="avatarRow-avatar avatarRow-avatar--count"
    >
      <span
        className="typemod--book typemod--small"
      >
        +
        2
      </span>
    </WithStyles(Avatar)>
  </li>
</ul>
`;
